src/extractor/guidance/intersection_handler.cpp

```diff
@@ -383,131 +394,147 @@ std::size_t IntersectionHandler::findObviousTurn(const EdgeID via_edge,
 
-        // don't prefer low priority classes
-        if (best != 0 && out_data.road_classification.IsLowPriorityRoadClass() &&
-            !current_best_class.IsLowPriorityRoadClass())
-            continue;
+        out.deviation_from_straight = angularDeviation(intersection[i].angle, STRAIGHT_ANGLE);
+        out.same_or_higher_priority =
+            out_node.road_classification.GetPriority() <= in_way.road_classification.GetPriority();
+        out.same_classification = out_classification == in_way.road_classification;
+        out.same_name_id = out_node.name_id == in_way.name_id;
 
-        const bool is_better_choice_by_priority =
-            best == 0 || obviousByRoadClass(in_data.road_classification,
-                                            out_data.road_classification,
-                                            current_best_class);
+        out_ways.push_back(out);
+    }
 
-        const bool other_is_better_choice_by_priority =
-            best != 0 && obviousByRoadClass(in_data.road_classification,
-                                            current_best_class,
-                                            out_data.road_classification);
+    const auto End = end(out_ways);
+    // dead end, or otherwise no exitable ways
+    if (num_valid_entry == 0)
+        return 0;
 
-        if ((!other_is_better_choice_by_priority && deviation < best_deviation) ||
-            is_better_choice_by_priority)
-        {
-            best_deviation = deviation;
-            best = i;
-        }
+    // sort out ways by what they share with the in way
+    // entry allowed, lowest deviation to greatest deviation, name, classification and priority
+    const auto sort_by = [](const out_way &lhs, const out_way &rhs) {
+        auto left_tie =
+            std::tie(lhs.same_classification, lhs.same_or_higher_priority);
+        auto right_tie =
+            std::tie(rhs.same_classification, rhs.same_or_higher_priority);
+        return lhs.entry_allowed > rhs.entry_allowed ||
+               (lhs.entry_allowed == rhs.entry_allowed &&
+                (lhs.deviation_from_straight < rhs.deviation_from_straight ||
+                 (lhs.deviation_from_straight == rhs.deviation_from_straight &&
+                  left_tie > right_tie)));
+    };
+    std::stable_sort(begin(out_ways), end(out_ways), sort_by);
+
+    // best candidate is the first way sorted to have the smallest deviation from
+    // straight and is same or higher priority class
+    const auto sameOrHigher = [&](const out_way &way) {
+        return way.entry_allowed && way.same_or_higher_priority;
+    };
+
+    const auto notLowPriority = [&](const out_way &way) {
+        return way.entry_allowed &&
+               !node_based_graph.GetEdgeData(way.road->eid)
+                    .road_classification.IsLowPriorityRoadClass();
+    };
+    auto best_candidate_it = std::find_if(begin(out_ways), End, sameOrHigher);
```

Comment:
isn't this was the sorting is for? To make sure the best candidate is at the front?